File: ios/Bonfire/Mobile/Chat/FirebaseChatDatabase.swift
import Foundation
import RxSwift
import Firebase

class FirebaseChatDatabase: ChatDatabase {

    private let messagesDB: FIRDatabaseReference

    init(messagesDB: FIRDatabaseReference) {
        self.messagesDB = messagesDB
    }

    func chat(channel: Channel) -> Observable<Chat> {
        return messages(forChannel: channel)
            .rx_readValue()
            .map(toChat(forChannel: channel))
    }

    func sendMessage(message: Message, channel: Channel) {
        messages(forChannel: channel).childByAutoId().setValue(message.asFirebaseValue())
    }

    private func messages(forChannel channel: Channel) -> FIRDatabaseReference {
        return messagesDB.child(channel.name)
    }

    private func toChat(forChannel channel: Channel) -> (FIRDataSnapshot throws -> Chat) {
        return { snapshot in
            let firebaseMessages = snapshot.children.allObjects
            let messages = try firebaseMessages.map {$0.value}.map(Message.init)
            return Chat(channel: channel, messages: messages)
        }
    }
}

File: android/app/src/main/java/com/novoda/bonfire/chat/database/FirebaseChatDatabase.java
package com.novoda.bonfire.chat.database;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.novoda.bonfire.channel.data.model.Channel;
import com.novoda.bonfire.chat.data.model.Chat;
import com.novoda.bonfire.chat.data.model.Message;
import com.novoda.bonfire.rx.FirebaseObservableListeners;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

public class FirebaseChatDatabase implements ChatDatabase {

    private static final int DEFAULT_LIMIT = 1000;

    private final DatabaseReference messagesDB;
    private final FirebaseObservableListeners firebaseObservableListeners;

    public FirebaseChatDatabase(FirebaseDatabase firebaseDatabase, FirebaseObservableListeners firebaseObservableListeners) {
        messagesDB = firebaseDatabase.getReference("messages");
        this.firebaseObservableListeners = firebaseObservableListeners;
    }

    @Override
    public Observable<Chat> observeChat(Channel channel) {
        return firebaseObservableListeners.listenToValueEvents(messagesInChannel(channel).limitToLast(DEFAULT_LIMIT), toChat());
    }

    @Override
    public void sendMessage(Channel channel, Message message) {
        messagesInChannel(channel).push().setValue(message); //TODO handle errors
    }

    private DatabaseReference messagesInChannel(Channel channel) {
        return messagesDB.child(channel.getName());
    }

    private Function<DataSnapshot, Chat> toChat() {
        return new Function<DataSnapshot, Chat>() {
            @Override
            public Chat apply(@NonNull DataSnapshot dataSnapshot) throws Exception {
                Iterable<DataSnapshot> children = dataSnapshot.getChildren();
                List<Message> messages = new ArrayList<>();
                for (DataSnapshot child : children) {
                    Message message = child.getValue(Message.class);
                    messages.add(message);
                }
                return new Chat(messages);
            }
        };
    }

}
